modules/targetanalysis/src/pixel_object_list.cpp

```diff
@@ -0,0 +1,332 @@
+
+using namespace boost;
+using namespace std;
+using namespace cv;
+
+
+PixelObjectList* PixelObjectList::firstInstance = NULL;
+
+PixelObjectList::~PixelObjectList(){
+    int i = 0;
+    poNode* tempPointer;
+    while(i<listLength){
+        tempPointer = tail;
+        tail = tail->next;
+        delete(tempPointer);
+        i++;
+    }
+}
+
+void PixelObjectList::addNode(PixelObject* po){
+    //Initialize node
+    struct poNode* newNode = new struct poNode;
+    //Initialize object within node
+    Object* newObject = new Object;
+    newNode->o = newObject;
+        
+    newNode->o->add_pobject(po);
+    newNode->o->update();
+    newNode->next = 0; //Nullify pointer (Since there is no next list item)
+
+    //Update old node with the new node
+    if (head){  //If HEAD has been initialized already
+        head->next = newNode;
+    }
+    else{ //If HEAD hasn't been initialized already
+        tail = newNode;
+    }
+    //Change head to represent the new node
+    head = newNode;
+    //Update list length
+    listLength++;      
+    BOOST_LOG_TRIVIAL(debug) << "Node added. New List Length: " << listLength;
+}
+
+double PixelObjectList::compareNode(PixelObject* po1, Object* o2){
+    //Use FLANN here?
+    //Compare Pixel Objects based on Metadata and visual queues
+    //Uses Hierarchical Structure: GPS data is most important, followed by
+    //Contour/Shape, followed by Colour
+    TargetAnalyzer* ta = TargetAnalyzer::getInstance();
+    using TA = TargetAnalyzer; 
+    double gps = compareGPS(po1, o2);
+    if (gps > ta->get_threshold(TA::GPS)){
+        double visual = compareContours(po1, o2);
+        if (visual > ta->get_threshold(TA::CONTOUR)){
+            double colour = compareColour(po1, o2);
+            if (colour > ta->get_threshold(TA::COLOUR)){
+                return (gps + ta->get_threshold_bias(TA::GPS))*(visual +
+                ta->get_threshold_bias(TA::CONTOUR))*(colour + ta->get_threshold_bias(TA::COLOUR)); 
+            }
+        }
+    }
+    return 0;
+}
+
+double PixelObjectList::compareColour(PixelObject* po1, Object* o2){
+    cv::Scalar a = po1->get_colour();
+    cv::Scalar b = o2->get_colour();
+    double* aArr = a.val;
+    double* bArr = b.val;
+
+    double euclideanDistance = sqrt(pow(aArr[0] - bArr[0],2) + pow(aArr[1] -
+    bArr[1],2) + pow(aArr[2] - bArr[2],2));
+
+    return 1 - euclideanDistance/sqrt(3*pow(255,2));
+}
+
+double PixelObjectList::compareGPS(PixelObject* po1, Object* o2){
+    //Calculations assume ideal scenario (no tilt in photos, no distortion)
+    Frame* f = po1->get_image();
+    cv::Point2d result(0,0);
+    if(!getGPS(po1->get_centroid(),cv::Point2d(69.5/2,125.3/2)/*getSettings()*/,f,&result)){
+        //If there is something wrong with the GPS coordinates, its not a valid
+        //match
+        return 0;
+    }
+    po1->set_gps_centroid(result);
+    double distance = TargetAnalyzer::getInstance()->getGPSDistance(result, o2->get_centroid());
```

Comment:
Agreed. A better solution would be to store all GPS coordinates and then use K-Means clustering....probably won't happen before competition.